### ISSUE TITLE
Label not set: failure-domain.beta.kubernetes.io/zone

### ISSUE THREAD

**Issue by xmik** [kubernetes/cloud-provider-openstack#191]
/kind question

## TL;DR: Problem
This label is not being set on any node in my Kubernetes cluster:
```
failure-domain.beta.kubernetes.io/zone=nova
```
In result, a cinder volume (created from PVC) is created but not attached to any vm. 
This label should be set on a node by cloudprovider, [source](https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#failure-domain-beta-kubernetes-io-zone).
Error from Pod waiting for this volume:
```
$ kubectl -n testing describe pod
Name:           nginx-vols-testing-c75cbd664-5m7qq
Namespace:      testing
# ...
Tolerations:     node.cloudprovider.kubernetes.io/uninitialized
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                 From               Message
  ----     ------            ----                ----               -------
  Warning  FailedScheduling  25s (x322 over 1h)  default-scheduler  0/3 nodes are available: 1 node(s) didn't match node selector, 2 node(s) had no available volume zone.
```

Setting the label manually with:
```
kubectl label node/<node-name> failure-domain.beta.kubernetes.io/zone=nova
```
fixes it (volume becomes attached to a VM).

## Environment
I have deployed Kubernetes 1.10 cluster on OpenStack VMs, intend to use cinder for persistent volumes. Now, I am not sure which solution should I choose and I have trouble understanding [openstack-kubernetes-integration-options.md](https://github.com/kubernetes/cloud-provider-openstack/blob/v0.1.0/docs/openstack-kubernetes-integration-options.md).

Can you please correct me if I am wrong? The solutions (I think) are:
1. **In-tree OpenStack provider in Kubernetes repository**, the old way: use only kube-controller-manager (no new ccm containers needed). The flags to set for kubelet, kube-apiserver and kube-controller-manager: `--cloud-config=/path/to/cloud_config  --cloud-provider=openstack`
2. **Cloud Controller Manager (CCM) in Kubernetes repository**, the new, but temporary way: 
   * use kube-controller-manager, but set `--external-cloud-volume-plugin=openstack`, so that it has a role of: Volume controller, [source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#persistentvolumelabels-controller), but also set: `--cloud-config=/path/to/cloud_config  --cloud-provider=external`
   * and also use cloud-controller-manager with `--cloud-config=/path/to/cloud_config  --cloud-provider=openstack`. This is just one container from image `k8scloudprovider/openstack-cloud-controller-manager:v0.1.0`
3. **External OpenStack provider**, which is the same as the 2nd way, but we should also deploy one from those 3:
   * Cinder Standalone provisioner
   * Cinder Flex volume driver
   * Cinder CSI driver

My goal is to use Cinder CSI driver, because:
   * "CSI is planned to become the primary volume plugin system for Kubernetes"  [source](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md)
   * In the long run we should prefer CSI over FlexVolume [source](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md#working-with-out-of-tree-volume-plugin-options)
   * AFAIK Cinder Standalone provisioner is for those Kubernetes clusters that want to connect with external Cinder resources (and such a Kubernetes cluster doesn't need to be deployed on Openstack VMs).

But, since the CSI driver is WIP, [source](https://github.com/kubernetes/cloud-provider-openstack/blob/v0.1.0/docs/openstack-kubernetes-integration-options.md#cinder-csi-driver), I choose the 2nd way for now.

**Comment by jejer** [comment]
and ignore-volume-az not working.
https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#openstack

**Comment by lingxiankong** [comment]
should be fixed by #405, @xmik could you please confirm before i close this issue?

**Comment by xmik** [comment]
@lingxiankong, I will check that some day before christmas.

**Comment by xmik** [comment]
@lingxiankong, I confirm it is fixed. Tested with cloud-provider-openstack commit: cc82c7f119cd63e857c47c34796a272cc967cf02. Thanks! :)

**Comment by lingxiankong** [comment]
@xmik thanks for the feedback!

**Comment by xmik** [comment]
I'm closing this issue, because there is v1.13.1 released with commit from #405.

**Comment by alkimake** [comment]
I have the same problem with `openstack-cloud-controller-manager` v1.13.1. Still cloud controller manager does not set label `failure-domain.beta.kubernetes.io/zone=nova`

**Comment by lingxiankong** [comment]
@alkimake did you see any error in openstack-cloud-controller-manager log?

**Comment by alkimake** [comment]
I had a big problem about logs, which is caused by kubernetes static pod problems. Whenever i change configurations, kubernetes gives me previous pod logs. 
Basically i upgraded cloud-controller-manager from v0.2.0, which had the same problem. I upgraded and saw no errors. But i have seen that  `region` is set properly. I wonder may be it is an issue because of upgrade. 
So i am gonna deploy a new stack with latest version of controller-manager, and see what happens. I will inform you later on.

**Comment by alkimake** [comment]
I think the error caused because of upgrade process from v0.2.0. I have setup a new cluster with same openstack credentials, it works perfectly. Lets ignore my comment for now.
